File: trwm/preflight.py
from __future__ import annotations

from dataclasses import dataclass
from math import isfinite
from typing import Iterable, Sequence


Matrix = tuple[tuple[float, ...], ...]
Vector = tuple[float, ...]


@dataclass(frozen=True)
class ShapeRankPreflight:
    r90: int
    fits_budget: bool
    energy_at_budget: float
    cumulative_energy: tuple[float, ...]
    component_energy: tuple[float, ...]
    total_energy: float


def shape_rank_preflight(
    target_updates: Iterable[Sequence[float]],
    output_map: Sequence[Sequence[float]] | None = None,
    rank_budget: int = 4,
    threshold: float = 0.90,
) -> ShapeRankPreflight:
    updates = tuple(tuple(float(value) for value in update) for update in target_updates)
    if not updates:
        raise ValueError("target_updates must not be empty")
    width = len(updates[0])
    if width == 0:
        raise ValueError("target update vectors must not be empty")
    if any(len(update) != width for update in updates):
        raise ValueError("all target update vectors must have the same length")
    if any(not isfinite(value) for update in updates for value in update):
        raise ValueError("target update values must be finite")
    if not isinstance(rank_budget, int) or isinstance(rank_budget, bool) or rank_budget < 0:
        raise ValueError("rank_budget must be a non-negative integer")
    if not isfinite(threshold) or not 0 < threshold <= 1:
        raise ValueError("threshold must be in (0, 1]")

    matrix = _normalize_matrix(output_map, width)
    eigenvalues, eigenvectors = _jacobi_eigen_symmetric(_gram(matrix))
    components = sorted(
        (
            max(0.0, eigenvalue)
            * sum(_dot(vector, update) ** 2 for update in updates),
            idx,
        )
        for idx, (eigenvalue, vector) in enumerate(zip(eigenvalues, eigenvectors))
    )
    component_energy = tuple(value for value, _idx in sorted(components, key=lambda row: (-row[0], row[1])))
    total = sum(component_energy)
    if total <= 0:
        cumulative = tuple(0.0 for _ in component_energy)
        r90 = 0
        energy_at_budget = 0.0
    else:
        running = 0.0
        cumulative_values = []
        r90 = len(component_energy)
        for idx, value in enumerate(component_energy, start=1):
            running += value
            ratio = running / total
            cumulative_values.append(ratio)
            if ratio >= threshold and r90 == len(component_energy):
                r90 = idx
        cumulative = tuple(cumulative_values)
        energy_at_budget = cumulative[min(rank_budget, len(cumulative)) - 1] if rank_budget > 0 and cumulative else 0.0
    return ShapeRankPreflight(
        r90=r90,
        fits_budget=r90 <= rank_budget,
        energy_at_budget=energy_at_budget,
        cumulative_energy=cumulative,
        component_energy=component_energy,
        total_energy=total,
    )


def one_hot_updates(labels: Iterable[int], label_count: int) -> tuple[Vector, ...]:
    if label_count <= 0:
        raise ValueError("label_count must be positive")
    updates = []
    for label in labels:
        if not 0 <= label < label_count:
            raise ValueError(f"label outside range: {label}")
        updates.append(tuple(1.0 if idx == label else 0.0 for idx in range(label_count)))
    return tuple(updates)


def identity_output_map(width: int) -> Matrix:
    if width <= 0:
        raise ValueError("width must be positive")
    return tuple(tuple(1.0 if row == col else 0.0 for col in range(width)) for row in range(width))


def _normalize_matrix(output_map: Sequence[Sequence[float]] | None, width: int) -> Matrix:
    if output_map is None:
        return identity_output_map(width)
    matrix = tuple(tuple(float(value) for value in row) for row in output_map)
    if not matrix:
        raise ValueError("output_map must not be empty")
    if any(len(row) != width for row in matrix):
        raise ValueError("output_map column count must match target update width")
    if any(not isfinite(value) for row in matrix for value in row):
        raise ValueError("output_map values must be finite")
    return matrix


def _gram(matrix: Matrix) -> Matrix:
    cols = len(matrix[0])
    return tuple(
        tuple(sum(row[i] * row[j] for row in matrix) for j in range(cols))
        for i in range(cols)
    )


def _jacobi_eigen_symmetric(matrix: Matrix, max_sweeps: int = 80, tolerance: float = 1e-12) -> tuple[Vector, tuple[Vector, ...]]:
    n = len(matrix)
    if n == 0 or any(len(row) != n for row in matrix):
        raise ValueError("matrix must be square")
    a = [list(row) for row in matrix]
    vectors = [[1.0 if row == col else 0.0 for col in range(n)] for row in range(n)]
    for _sweep in range(max_sweeps):
        p, q, max_value = 0, 1 if n > 1 else 0, 0.0
        for i in range(n):
            for j in range(i + 1, n):
                value = abs(a[i][j])
                if value > max_value:
                    p, q, max_value = i, j, value
        if max_value < tolerance:
            break
        if a[p][p] == a[q][q]:
            angle = 0.7853981633974483
        else:
            angle = 0.5 * _atan2(2.0 * a[p][q], a[q][q] - a[p][p])
        c = _cos(angle)
        s = _sin(angle)
        app = c * c * a[p][p] - 2.0 * s * c * a[p][q] + s * s * a[q][q]
        aqq = s * s * a[p][p] + 2.0 * s * c * a[p][q] + c * c * a[q][q]
        a[p][p] = app
        a[q][q] = aqq
        a[p][q] = 0.0
        a[q][p] = 0.0
        for k in range(n):
            if k in {p, q}:
                continue
            akp = c * a[k][p] - s * a[k][q]
            akq = s * a[k][p] + c * a[k][q]
            a[k][p] = a[p][k] = akp
            a[k][q] = a[q][k] = akq
        for k in range(n):
            vkp = c * vectors[k][p] - s * vectors[k][q]
            vkq = s * vectors[k][p] + c * vectors[k][q]
            vectors[k][p] = vkp
            vectors[k][q] = vkq
    values = tuple(a[i][i] for i in range(n))
    columns = tuple(tuple(vectors[row][col] for row in range(n)) for col in range(n))
    return values, columns


def _dot(a: Sequence[float], b: Sequence[float]) -> float:
    return sum(x * y for x, y in zip(a, b))


def _atan2(y: float, x: float) -> float:
    # Local import keeps the public dependency footprint obvious.
    from math import atan2

    return atan2(y, x)


def _cos(x: float) -> float:
    from math import cos

    return cos(x)


def _sin(x: float) -> float:
    from math import sin

    return sin(x)

File: src/preflight.ts
export type NumericVector = number[];
export type NumericMatrix = number[][];

export interface ShapeRankPreflight {
  r90: number;
  fitsBudget: boolean;
  energyAtBudget: number;
  cumulativeEnergy: number[];
  componentEnergy: number[];
  totalEnergy: number;
}

export function shapeRankPreflight(
  targetUpdates: NumericVector[],
  outputMap?: NumericMatrix,
  rankBudget = 4,
  threshold = 0.9,
): ShapeRankPreflight {
  if (targetUpdates.length === 0) {
    throw new RangeError("targetUpdates must not be empty");
  }
  const width = targetUpdates[0].length;
  if (width === 0) {
    throw new RangeError("target update vectors must not be empty");
  }
  if (targetUpdates.some((update) => update.length !== width)) {
    throw new RangeError("all target update vectors must have the same length");
  }
  if (targetUpdates.some((update) => update.some((value) => !Number.isFinite(value)))) {
    throw new TypeError("target update values must be finite");
  }
  if (!Number.isInteger(rankBudget) || rankBudget < 0) {
    throw new RangeError("rankBudget must be a non-negative integer");
  }
  if (!Number.isFinite(threshold) || threshold <= 0 || threshold > 1) {
    throw new RangeError("threshold must be in (0, 1]");
  }
  const matrix = normalizeMatrix(outputMap, width);
  const { values, vectors } = jacobiEigenSymmetric(gram(matrix));
  const componentEnergy = values
    .map((value, idx) => Math.max(0, value) * targetUpdates.reduce((sum, update) => {
      const coeff = dot(vectors[idx], update);
      return sum + coeff * coeff;
    }, 0))
    .sort((a, b) => b - a);
  const totalEnergy = componentEnergy.reduce((sum, value) => sum + value, 0);
  let r90 = 0;
  let energyAtBudget = 0;
  const cumulativeEnergy: number[] = [];
  if (totalEnergy > 0) {
    let running = 0;
    r90 = componentEnergy.length;
    for (let idx = 0; idx < componentEnergy.length; idx += 1) {
      running += componentEnergy[idx];
      const ratio = running / totalEnergy;
      cumulativeEnergy.push(ratio);
      if (ratio >= threshold && r90 === componentEnergy.length) {
        r90 = idx + 1;
      }
    }
    energyAtBudget = rankBudget > 0 && cumulativeEnergy.length > 0
      ? cumulativeEnergy[Math.min(rankBudget, cumulativeEnergy.length) - 1]
      : 0;
  }
  return {
    r90,
    fitsBudget: r90 <= rankBudget,
    energyAtBudget,
    cumulativeEnergy,
    componentEnergy,
    totalEnergy,
  };
}

export function oneHotUpdates(labels: number[], labelCount: number): NumericVector[] {
  if (!Number.isInteger(labelCount) || labelCount <= 0) {
    throw new RangeError("labelCount must be a positive integer");
  }
  return labels.map((label) => {
    if (!Number.isInteger(label) || label < 0 || label >= labelCount) {
      throw new RangeError(`label outside range: ${label}`);
    }
    return Array.from({ length: labelCount }, (_unused, idx) => idx === label ? 1 : 0);
  });
}

export function identityOutputMap(width: number): NumericMatrix {
  if (!Number.isInteger(width) || width <= 0) {
    throw new RangeError("width must be a positive integer");
  }
  return Array.from({ length: width }, (_unused, row) =>
    Array.from({ length: width }, (_unusedCol, col) => row === col ? 1 : 0),
  );
}

function normalizeMatrix(outputMap: NumericMatrix | undefined, width: number): NumericMatrix {
  if (!outputMap) {
    return identityOutputMap(width);
  }
  if (outputMap.length === 0 || outputMap.some((row) => row.length !== width)) {
    throw new RangeError("outputMap column count must match target update width");
  }
  return outputMap.map((row) => row.map((value) => {
    if (!Number.isFinite(value)) {
      throw new TypeError("outputMap values must be finite");
    }
    return value;
  }));
}

function gram(matrix: NumericMatrix): NumericMatrix {
  const cols = matrix[0].length;
  return Array.from({ length: cols }, (_unused, i) =>
    Array.from({ length: cols }, (_unusedCol, j) =>
      matrix.reduce((sum, row) => sum + row[i] * row[j], 0),
    ),
  );
}

function jacobiEigenSymmetric(matrix: NumericMatrix, maxSweeps = 80, tolerance = 1e-12): { values: number[]; vectors: NumericVector[] } {
  const n = matrix.length;
  if (n === 0 || matrix.some((row) => row.length !== n)) {
    throw new RangeError("matrix must be square");
  }
  const a = matrix.map((row) => [...row]);
  const basis = Array.from({ length: n }, (_unused, row) =>
    Array.from({ length: n }, (_unusedCol, col) => row === col ? 1 : 0),
  );
  for (let sweep = 0; sweep < maxSweeps; sweep += 1) {
    let p = 0;
    let q = n > 1 ? 1 : 0;
    let maxValue = 0;
    for (let i = 0; i < n; i += 1) {
      for (let j = i + 1; j < n; j += 1) {
        const value = Math.abs(a[i][j]);
        if (value > maxValue) {
          p = i;
          q = j;
          maxValue = value;
        }
      }
    }
    if (maxValue < tolerance) break;
    const angle = a[p][p] === a[q][q]
      ? Math.PI / 4
      : 0.5 * Math.atan2(2 * a[p][q], a[q][q] - a[p][p]);
    const c = Math.cos(angle);
    const s = Math.sin(angle);
    const app = c * c * a[p][p] - 2 * s * c * a[p][q] + s * s * a[q][q];
    const aqq = s * s * a[p][p] + 2 * s * c * a[p][q] + c * c * a[q][q];
    a[p][p] = app;
    a[q][q] = aqq;
    a[p][q] = 0;
    a[q][p] = 0;
    for (let k = 0; k < n; k += 1) {
      if (k === p || k === q) continue;
      const akp = c * a[k][p] - s * a[k][q];
      const akq = s * a[k][p] + c * a[k][q];
      a[k][p] = akp;
      a[p][k] = akp;
      a[k][q] = akq;
      a[q][k] = akq;
    }
    for (let k = 0; k < n; k += 1) {
      const vkp = c * basis[k][p] - s * basis[k][q];
      const vkq = s * basis[k][p] + c * basis[k][q];
      basis[k][p] = vkp;
      basis[k][q] = vkq;
    }
  }
  return {
    values: Array.from({ length: n }, (_unused, idx) => a[idx][idx]),
    vectors: Array.from({ length: n }, (_unused, col) =>
      Array.from({ length: n }, (_unusedRow, row) => basis[row][col]),
    ),
  };
}

function dot(a: NumericVector, b: NumericVector): number {
  return a.reduce((sum, value, idx) => sum + value * b[idx], 0);
}
